src/charm.py

```diff
@@ -1,104 +1,432 @@
+        """
+        # Check gitlab-related options:
+        gitlab_project_visibility = GITLAB_PROJECT_VISIBILITY_PRIVATE
+        if self.model.config['gitlab-create-new-projects-as-public']:
+            gitlab_project_visibility = GITLAB_PROJECT_VISIBILITY_PUBLIC
+        # TODO(aznashwan): remove this check on eventual Gitlab relation:
+        gitlab_client_id = self.model.config.get('gitlab-client-id')
+        gitlab_client_secret = self.model.config.get('gitlab-client-secret')
+        gitlab_project_tag = self.model.config['gitlab-project-tag']
+        gitlab_project_creation_group_pattern = (
+            self.model.config['gitlab-project-creation-group-pattern'])
+        if not all([
+                gitlab_project_visibility, gitlab_client_id,
+                gitlab_client_secret, gitlab_project_tag,
+                gitlab_project_creation_group_pattern]):
+            return model.BlockedStatus(
+                "One or more Gitlab-related charm configuration options "
+                "are missing.")
+
+        # Check Java logging options:
+        request_logging_level = self._get_logging_level_from_config(
+            "server-requests-logging-level")
+        server_logging_level = self._get_logging_level_from_config(
+            "server-logging-level")
+        server_logging_format = self.model.config['server-logging-format']
+        if not all([server_logging_level, request_logging_level]):
+            return model.BlockedStatus(
+                "One or more logging config options are improperly formatted "
+                "or missing. Please review the debug-log for more details.")
+
+        # Check Mongo-related options:
+        mongo_creds = self._stored.mongodb_credentials
+        if not mongo_creds or 'replica_set_uri' not in mongo_creds:
+            return model.BlockedStatus(
+                "No stored MongoDB credentials were found yet. Please "
+                "ensure the Charm is properly related to MongoDB.")
+        mongo_replica_set_uri = self._stored.mongodb_credentials[
+            'replica_set_uri']
+        databases = mongo_creds.get('databases')
+        database_name = None
+        if databases:
+            database_name = databases[0]
+            # NOTE(aznashwan): the Java MongoDB can't handle DB names in the
+            # URL, so we need to trim that part and pass the database name
+            # as a separate parameter within the config as the
+            # sdlc_config['pac4j']['mongoDb'] option below.
+            split_uri = [
+                elem
+                for elem in mongo_replica_set_uri.split('/')[:-1]
+                # NOTE: filter any empty strings resulting from double-slashes:
+                if elem]
+            # NOTE: schema prefix needs two slashes added back:
+            mongo_replica_set_uri = "%s//%s" % (
+                split_uri[0], "/".join(split_uri[1:]))
+
+        # Compile base config:
+        sdlc_config.update({
```

Comment:
This is a pretty large blob of stuff for the charm code, you could consider modelling the legend config as a dataclass in a separate file, that takes `self.config` / `databases` in the constructor and implements `to_dict()` on the class to return a dictionary representation?